File: src/python/core/src/dicebear/style_def/meta.py
"""Lazily-constructed view over a style definition's ``meta`` block."""

from __future__ import annotations

from typing import Any

from .meta_creator import MetaCreator
from .meta_license import MetaLicense
from .meta_source import MetaSource


class Meta:
    """Lazily-constructed view over a style definition's ``meta`` block.

    Exposes the license, creator, and source descriptors.
    """

    def __init__(self, data: dict[str, Any]) -> None:
        self._data = data
        self._license: MetaLicense | None = None
        self._creator: MetaCreator | None = None
        self._source: MetaSource | None = None

    def license(self) -> MetaLicense:
        """Return the license descriptor, defaulting to an empty object."""
        if self._license is None:
            self._license = MetaLicense(self._data.get("license", {}))

        return self._license

    def creator(self) -> MetaCreator:
        """Return the creator descriptor, defaulting to an empty object."""
        if self._creator is None:
            self._creator = MetaCreator(self._data.get("creator", {}))

        return self._creator

    def source(self) -> MetaSource:
        """Return the source descriptor, defaulting to an empty object."""
        if self._source is None:
            self._source = MetaSource(self._data.get("source", {}))

        return self._source

File: src/csharp/core/src/Style/Meta.cs
using System.Text.Json.Nodes;

namespace DiceBear.Internal
{
    /// <summary>
    /// Lazily-constructed view over a style definition's <c>meta</c> block,
    /// exposing the license, creator and source descriptors.
    /// </summary>
    internal sealed class Meta
    {
        private readonly JsonObject _data;

        private MetaLicense? _license;
        private MetaCreator? _creator;
        private MetaSource? _source;

        internal Meta(JsonObject data) => _data = data;

        /// <summary>
        /// Returns the license descriptor, defaulting to an empty object when
        /// the style definition omits the field.
        /// </summary>
        internal MetaLicense License() =>
            _license ??= new MetaLicense(JsonRead.Obj(_data, "license") ?? new JsonObject());

        /// <summary>
        /// Returns the creator descriptor, defaulting to an empty object when
        /// the style definition omits the field.
        /// </summary>
        internal MetaCreator Creator() =>
            _creator ??= new MetaCreator(JsonRead.Obj(_data, "creator") ?? new JsonObject());

        /// <summary>
        /// Returns the source descriptor, defaulting to an empty object when
        /// the style definition omits the field.
        /// </summary>
        internal MetaSource Source() =>
            _source ??= new MetaSource(JsonRead.Obj(_data, "source") ?? new JsonObject());
    }
}
